usr/lib/common/pkey_utils.c

```diff
@@ -908,18 +908,50 @@ struct aes_xts_param {
     uint8_t param_pcc[sizeof(struct cpacf_pcc_xts_aes_256_param)];
     unsigned int fc;
     unsigned int keylen;
+    convert_key_t convert_key;
+    STDLL_TokData_t *tokdata;
+    SESSION *session;
+    OBJECT *key_obj;
 };
 
 static CK_RV pkey_aes_xts_iv_from_tweak(CK_BYTE *tweak, CK_BYTE* iv,
                                         void *cb_data)
 {
     struct aes_xts_param *param = cb_data;
-    int offset, rc;
+    convert_key_t convert_key = param->convert_key;
+    int offset, rc = 1, num_tries = 0;
+    CK_BYTE protkey[MAXPROTKEYSIZE * 2];
+    CK_ULONG protkey_len = sizeof(protkey);
+    CK_RV ret;
 
     offset = AES_XTS_PCC_I(param->keylen * 8);
     memcpy(param->param_pcc + offset, tweak, AES_BLOCK_SIZE);
 
-    rc = s390_pcc(param->fc & 0x7f, param->param_pcc);
+    while (rc != 0 && num_tries < PKEY_CONVERT_KEY_RETRIES) {
+        rc = s390_pcc(param->fc & 0x7f, param->param_pcc);
+        if (rc != 0) {
+            TRACE_DEVEL("%s rc from s390_pcc = %d, probably caused by "
+                        "an LGR. Retrying ...\n", __func__, rc);
+            ret = convert_key(param->tokdata, param->session, param->key_obj,
+                              CK_TRUE, protkey, &protkey_len);
+            if (ret != CKR_OK)
+                break;
+            /*
+             * Copy XTS key 1 into CPACF parmblock for KM instruction. After
+             * re-deriving the iv from tweak, the subsequent calls to
+             * pkey_aes_xts_cipher_blocks will then use the re-derived key
+             * as well.
+             */
+            memcpy(param->param_km, protkey, protkey_len / 2);
```

Comment:
There are many places in the PKEY code where the protected key is copied to a CPACF param block without any length checks. While it is very unlikely that `protkey_len` is larger than `param_km`, it is not impossible. `param_km` is 64+16 bytes, but the protkey buffer is MAXPROTKEYSIZE+2 in size, which is most likely larger than 64+16.

This does not only affect this place, but many places. Most of the time the protected key is copied right from the attribute, e.g. like this:
`memcpy(param.key, pkey_attr->pValue, pkey_attr->ulValueLen);`
No check if `pkey_attr->ulValueLen` is larger than the target of the copy.....

Since attributes can theoretically be set/updated by a malicious user program with data of any size (i.e. larger size), just copying this without a length check can cause memory overlays, etc.  

We should add explicit length checks at all these places, or at least do something like this to prevent copying too many bytes:

`memcpy(param.key, pkey_attr->pValue, MIN(pkey_attr->ulValueLen, sizeof(param.key)));`

Having a too sort or otherwise invalid protected key in the parm block will cause a failure later, and a possible rederive of the prot key. So this will be handled anyway.  

Please do so in the commits if this PR where such memcpy's are added, and add a new commit on top to fix all the other places not anyway touched by the commits of this PR.

Reply:
Done.